### ISSUE TITLE
Controls for editing text in Tables in R-Instat

### ISSUE THREAD

**Issue by lilyclements** [africanmathsinitiative/R-Instat#8554]
@rdstern and I have been discussing the tables and come up with a new set of controls for the Titles tab when editing the subtitle.

@N-thony we wanted to know if the following is possible, and, if you had any additional requests or ideas for this?

We want to have a control that is similar to this:

<img width="234" alt="image" src="https://github.com/africanmathsinitiative/R-Instat/assets/21180424/d32574ff-5de7-4ff4-9b90-66dc975c3c54">

But with the options given in the `gt` package in R, so instead we would have something like this:

![image](https://github.com/africanmathsinitiative/R-Instat/assets/21180424/d1513706-5589-4fdd-b477-aab1718b192c)

I've numbered the different bits and go through the questions I have on this for @N-thony 

1. **Font** - In here, we have an editable drop down giving the font to choose. Is there already a control like this in Visual Basic that we can use?
2. **Size** - This is the size of the font, and this can be an editable drop down that allows numerical values only. We could have in it the options that Word has: 8, 9, 10, 11, 12, 14, 16, 18, 20, 22, 24, 26, 28, 36, 48, 72
3. **Bold** - This could be like a button, with an arrow in the corner to select further options (like we have on our OK) to add further bold options in R (e.g., _bolder, lighter_).
If you click the button, it makes it **bold**, if you unclick it, it removes that parameter and parameter.
Is this possible?
6. This is a colour changer to change the colour of your font. Is it possible to have this as an option in R-Instat as a button with an arrow of the different colours?

4-5, 7-9. Are the same idea as 3., but with **Italic**, **Underline**, **Sentence Case**, and **Alignment** There are additional options if you click on the little arrow in the corner that are offered in R. No additional questions here that isn't asked in 3.!

Let me know what you think @N-thony :)

**Comment by lilyclements** [comment]
_Sidenote on the R-Code Side for myself_
These are all in the `gt::tab_style` function. 
1. `font`
2. `size`
3. `weight`, options are: `"normal"`, `"bold"`, `"lighter"`, `"bolder"`. We run `"bold"` if the button is pressed, `"normal"` if not unpressed, and `"bold"`, `"lighter"`, and `"bolder"` are options on the drop down.
4. `style`, options are: `"normal"`, `"italic"`, `"oblique"`. We run `"italic"` if the button is pressed, `"normal"` if not unpressed, and `"italic"` and `"italic"` are options on the drop down.
5. `decorate` options are `"overline"`, `"line-through"`, or `"underline"`. We run `"underline"` if the button is pressed, `NULL` if not unpressed, and `"overline"`, `"line-through"`, or `"underline"` are options on the drop down.
6. `colour` the colour of the text.
7. `transform`, options are "uppercase", "lowercase", or "capitalize" on the drop down.
8. `align` - can be one of either `"center"`, `"left"`, `"right"`, or `"justify"` on the drop down.
9. `v_align` options are `"middle"`, `"top"`, or `"bottom"` on the drop down.

**Comment by lilyclements** [comment]
On reflection this might be more of a question for @Patowhiz. What are your thoughts on this @Patowhiz? :)

**Comment by Patowhiz** [comment]
@lilyclements this sounds like a good functionality, where will this control be placed?

**Comment by lilyclements** [comment]
@Patowhiz excellent! It will be used in the tables subdialogs multiple times; for example, the title, the subtitle, the footnotes, stub, source, table text, column names, etc.

**Comment by Patowhiz** [comment]
Then I presume the controls should be formatted/styled/arranged like the ggplot controls we have for the graph options sub dialog? @rdstern do you think consistency should matter in this case?

**Comment by lilyclements** [comment]
@Patowhiz can you explain where in the ggplot controls you are referring to? Is it like in the Title tab? I would suggest that actually we think about (in time) replacing the ggplot sub dialog with this sort of design too.

**Comment by Patowhiz** [comment]
@lilyclements yes I was referring to something like the Title tab. Notice it uses controls that are similar to other tabs like the Y-axis tab.

**Comment by Patowhiz** [comment]
@lilyclements I think it's very possible. WinForms should have a similar toolbar set of controls. Sounds like a good task for @N-thony.

**Comment by lilyclements** [comment]
@Patowhiz great.

@N-thony what are your thoughts on working on this this week? After more discussions with @rdstern we've decided it might make more sense to have this set of controls on a sub-dialog itself, like this:

<img width="322" alt="image" src="https://github.com/africanmathsinitiative/R-Instat/assets/21180424/600a5294-9fc0-413c-85da-51692a50b109">

In our version, it would look like this

<img width="309" alt="image" src="https://github.com/africanmathsinitiative/R-Instat/assets/21180424/31b47fde-6d6d-4756-b8ec-d3630a9a2930">

I'll explain this here:
1. A check box that says "Edit text". If checked, then a groupbox titled "Font" is visible
2. In the font groupbox there are eight controls
3. **Font** In here, we have an editable drop down giving the font to choose. This is the `font` parameter, with the value as the item in the editable drop-down. 
4. **Style** Options are: `"normal", "lighter", "bold", "bolder"`, `"italic", "lighter-italic", "bold-italic", "bolder-italic"`, 
5. **Size**  This is the size of the font, and this can be an editable drop down that allows numerical values only. We could have in it the options that Word has: 8, 9, 10, 11, 12, 14, 16, 18, 20, 22, 24, 26, 28, 36, 48, 72.
6. **Font colour** How does the ucr Colour work? I assume it makes sense for this to be that ucr
7. **Underline style** Options are `"none"`, `"overline", "line-through", "underline"`.
8. **Text transform** Options are `default`, `"uppercase", "lowercase", "capitalize"`
9. **Align** Options are `default`, `"center", "left", "right", "justify"`
10. **Vertical align** Options are `default`, `"middle", "top", "bottom"`
11. A check box that says "Edit borders". If checked, then a groupbox titled "Borders/fill" is visible
12. In the borders/fill groupbox there are eight controls



@N-thony I'm told you're very good at design. Are you also into the coding side too (as in, should I mention the R functions?)

**Comment by lilyclements** [comment]
Great. This design is actually for two sub dialogs:

1. Format Text
2. Format Footnotes

They are pretty much the same code with a very small difference.

**Format Text**

R code:

```
list(
      gt::cell_text(),   # Added if "Edit text" is checked, contains the parameters in the first table
      gt::cell_fill(),     # Added if "Edit fill" is checked, contains the parameters in the second table, just the first one, "color".
      gt::cell_border()    # Added if "Edit border" is checked, contains the parameters in the second table, lines 2-5.
      )
```

**Format Footnotes**

R code:

```
  tab_style(
    style = list(
      gt::cell_text(),   # Added if "Edit text" is checked, contains the parameters in the first table
      gt::cell_fill(),     # Added if "Edit fill" is checked, contains the parameters in the second table, just the first one, "color".
      gt::cell_border()    # Added if "Edit border" is checked, contains the parameters in the second table, lines 2-5.
      ),
    locations = cells_footnotes()
  )
```

**Comment by rdstern** [comment]
@Patowhiz I am adding a "table" here as an example.  I also note there is a lot of information above and wonder if there are pointers there to something that can be used.  
I was using the simple amis data from the boot package, and wanting to check on it.  I quickly got 2 useful exploratory tables using the pivot-table dialog.  Here is one:

![image](https://github.com/IDEMSInternational/R-Instat/assets/11226469/1f3138b6-3b93-4557-b1ff-db4c364a80e1)

(This was useful for me to check on the "balance" of the data, namely the number of observations in each cell.)

and here is a second:

![image](https://github.com/IDEMSInternational/R-Instat/assets/11226469/75aa814f-612e-4839-b6d0-da5d2b977a82)

The odd ordering of the factor on the left, was because I had re-ordered into ascending order of the mean.  I wanted to check that it had done this.

However, I also present this, because there are many limitations of these tables (from the R-pivot-table package) compared to what I could do easily if I was only using a spreadsheet for the pivot tables.  I still hope that I can get "better" tables, atr least "static tables" from R-Instat than from a spreadsheet! 

(By the way I can now copy/paste these data into R-Instat, which is great, - except for the double-headed rows?  I have had that discussion also with the Genstat people, with their table stuff, many years ago.)

**Comment by rdstern** [comment]
@Patowhiz and @lilyclements and @jkmusyoka (copying Lily and James, in case they would like to add or change ideas below)  

I have been reflecting on our discussion yesterday.  In hindsight I am liking what you have done a lot.  I am not sure whether I have a "but", it is more an "and".  I would like to see whether we agree on each of my points below.
a) There are multiple graph dialogs.  I think we agreed there would be multiple Table dialogs?
b) There will also be a Tables sub-dialog with many tabs, just as there is a parallel graph sub-dialog.
c) It will be a single tables sub-dialog that is called by more than one main tables dialog.  So it will be pretty comprehensive, e.g. include a Titles tab.
d) The main tables dialogs will include important features, from the (single) sub-dialog, so some neat tables don't need to use the sub-dialog.
e) The first tables dialog (which you have been doing, will be the presentation of parts of an existing data frame.   This will also become the new Prepare > Data Frame > View Data dialog.  This reflects the fact that a data frame already looks like a table, in having both rows and columns.
f) It is possible (I hope) that a small extension to your existing dialog, could include nanoplots?  If I understand correctly, that would just be a new multiple receiver that includes the set of variables for the plot.  There would also be a pull-down for the type of plot and perhaps more.  If you agree, then I assume this could be a separate dialog, and this one would not be in the Prepare menu.
g) Then we do need another dialog.  It is what you called "advanced", but I suggest it will become the main tables dialog.   The main difference is in the "columns" receiver.  Remember there is a parallel between the graphs and the tables.  In a graph you start by asking what columns of data are for the y (the rows), and which are for the x (columns).  So, the dialog to adapt is  already almost there for you, namely:

(To continue)

**Comment by Patowhiz** [comment]
@rdstern I agree with your analysis above. From my deep dive of the `gt` package in that last few days, it looks like it is purely a presentation package. It expects you to do the margins and transformations then use it for presentation. This is different from pivot tables which have the ability to do the margins and advanced data transformations.

Previously I had the impression that it is capable of doing advanced data transformations but that is not the case. There are packages that extend it to compliment what it was not designed to do; **gtsummary**, **gtExtras**,  **pointblank**, **tfrmt**, **gto**. 

As a follow up, I also agree on the 3 stages that you proposed in the email:

> 1. Stage 1 is the summary stage from 13000 rows to an ordinary data frame with 51 rows.
> 2. Stage 2 is a simple reshaping of these summary data into a `gt` structure.  I don’t even want to consider this a data frame, it is a `gt` object. 
> 3. Stage 3 is the presentation of a `gt` object nicely.

Note, a `gt` object depends on a data frame. The starting point is always a data frame, what I mean is, you can't modify a `gt `object from a `gt` object (not unless I'm mistaken). So I think stage 2 and 3 will always start from a data frame.

So we can safely assume that the work I'm doing is purely on presentation tables. If you agree, we can split the tables issue into multiple issues to cover the different stages.

**Comment by lilyclements** [comment]
To add the spanner R code here as well as #8236 - 

```
library(gtsummary)
library(gt)

summarise_trial <- trial %>%
  dplyr::group_by(trt, grade, response) %>%
  dplyr::summarise(mean = mean(age, na.rm = TRUE))

pivot_summarise_trial <- pivot_wider(xx, names_from = c("trt", "grade"), values_from = mean)

pivot_summarise_trial %>%
  gt::gt() %>%
  gt::tab_spanner_delim(delim = "_") 
```

**Comment by lilyclements** [comment]
Closing this due to more recent conversatiosn on the tables that superseed this
